### PR TITLE
Fix issue #3568: allow sharing of activation function parameters along specified axes

### PR DESCRIPTION
**Pull request by the-moliver**
A general fix for all appropriate activation functions for this issue https://github.com/fchollet/keras/issues/3568

Now with tests included!

### CODE REVIEW

**Comment by the-moliver** [comment]
Bumping because this is important for replicating any convolutional models that use activation functions with free parameters.

**Comment by fchollet** [comment]
Hesitant to make any API changes at the moment. Also it feel like a bit of a niche use case, and one for which you could roll out your own layers. But I'll leave it open and consider later on.

**Comment by the-moliver** [comment]
I tried to make it only non-breaking additions that won't change how any existing code functions, but let me know what could be better. Also I don't think its really that niche, any use of parameterized nonlinearities applied to convolutional feature maps are subject to the original issue https://github.com/fchollet/keras/issues/3568 which is kind of a big deal I thought.

**Comment by singlasahil14** [comment]
PReLU is needed for many state of the art vision tasks.
https://github.com/fchollet/keras/issues/4452
In this very simple MNIST example, the first PReLU will have 18432 parameters when it should be 32.  I will have to use TF/caffe for my application because of this.

**Comment by fchollet** [comment]
Ok, agreed that we should add this feature. I'm wondering if we could avoid the use of `pattern_broadcast` which is a Theano-specific feature. A simple reshape wouldn't be inefficient since we'd only be adding unit dimensions. Also, `param_share_axes` isn't a very good name, the equivalent argument in Numpy is simply `axis`. We could consider `reduce_axes` or `reduction_axes`. Does this argument exist in other APIs out there, and what is it called?

**Comment by Rocketknight1** [comment]
I think in a lot of other APIs, there's no option for PReLU to have a tensor of slope parameters across more than one dimension, so your only choices are to have one shared slope parameter for all neurons in an entire layer, or to have N different slope parameters, where N is the dimension of one of the axes (usually the channels axis in a convnet, or the features axis for dense layers). I might be misinterpreting the Torch source though, so I could be wrong.

See: 
http://caffe.berkeleyvision.org/doxygen/classcaffe_1_1PReLULayer.html
https://github.com/torch/nn/blob/master/PReLU.lua

As for the parameter name, `reduction_axes` or possibly `unshared_axes`?

**Comment by the-moliver** [comment]
Thanks for the feedback. I wrote this to have the maximum flexibility in terms of sharing parameters across different axes, but I agree `param_share_axes` isn't the best name. I could change to `channels_shared` as @Rocketknight1 suggests. Thoughts?

To avoid having to keep track of, reshape and permute dimensions given this flexibility, broadcasting seemed like the right solution. I used pattern_broadcast because Theano doesn't automatically broadcast unit dimensions but Tensorflow does, so the new backend function only actually does something for Theano. It seemed cleanest to make the Keras broadcasting API behave like Theano, and it seemed useful to have broadcasting in the backend as well. Everything else seemed even more ugly. But I'm of course open to suggestions.

**Comment by eraly** [comment]
@the-moliver @fchollet 
I am re-writing the activation functions in dl4j (deep learning framework in java) to support such functionality and have been working on similar concepts. I came up with the term sharded dimension and would like to suggest the same.
Shared axis is technically not correct since when the alpha when per channel is shared across a subset of dimensions (rows and columns). Sharded seems more correct to me since the alpha is being duplicated across a dimension.

**Comment by the-moliver** [comment]
I'll leave it to our benevolent dictator to decide the best combination and order of one of {`channels`, `axes`, `dimensions`} with with one of {`shared`, `sharded`, `unshared`}. My feelings aren't particularly strong on this.

**Comment by fchollet** [comment]
I'd lean towards `reduction_axes` or `shared_axes`. As long as the docstring is explicit about what it does, either name seems fine to me. `reduction_axes` is probably the best.

File: keras/layers/advanced_activations.py [hunk 1]
```diff
@@ -73,7 +91,7 @@ def build(self, input_shape):
 
     def call(self, x, mask=None):
         pos = K.relu(x)
-        neg = self.alphas * (x - abs(x)) * 0.5
+        neg = K.pattern_broadcast(self.alphas, self.param_broadcast) * (x - abs(x)) * 0.5
```
Comment (fchollet) [inline keras/layers/advanced_activations.py]: I would rather have a Theano-only branch here `if K.backend() == 'theano'` than add a dummy method to the TF backend.

**Comment by the-moliver** [comment]
Made requested changes. I chose `shared_axes` because `reduction_axes` sounds to me like a reduction operation is being applied along certain axes, which is not the case.

**Comment by fchollet** [comment]
Tests are failing.

**Comment by the-moliver** [comment]
Tests updated to use `shared_axes` name.

**Comment by fchollet** [comment]
LGTM, thanks
